### PR TITLE
Add ManageComponent

### PR DESCRIPTION
**Pull request by ketal178**
# Add ManageComponent

## 介绍 / Description
Add ManageComponent
Add QFileShareToIpadActivity

## 更改内容 / Types of Changes
- 新功能 New Feature
- 增强优化 Enhancement/optimization

## 由这个PR解决或关闭的Issues / Issues Fixed or Closed by This PR

-  

## 检查列表 / Checklist
- [x] 我的代码符合本仓库的代码规范 My code follows the code style of this project
- [x] 我已经测试了这些更改，它们可以正常工作，既不会破坏原有任何功能(或我可以解决)，也不会影响对旧版本的支持 I have tested the changes and verified that they work and don't break anything(as well as I can manage) or drop the support for previous versions.
- [x] 我已合并了对后续工作无意义的commit并确认不会对后续维护造成困扰 I have merged commits that are meaningless for follow-up work and confirmed that they will not cause damage to follow-up maintenance

### CODE REVIEW

File: app/src/main/java/cc/ioctl/hook/JumpController.java [hunk 1]
```diff
@@ -66,7 +68,7 @@
     private ArrayList<Rule> rules = null;
 
     protected JumpController() {
-        super("qn_jmp_ctl_enable");
+        super("qn_jmp_ctl_enable", SyncUtils.PROC_MAIN, true);
```
Comment (NextAlone) [inline app/src/main/java/cc/ioctl/hook/JumpController.java:71]: true？

File: app/src/main/java/nil/nadph/qnotified/activity/ConfigV2Activity.java [hunk 2]
```diff
@@ -69,6 +75,14 @@ protected void onCreate(Bundle savedInstanceState) {
         if (R.string.res_inject_success >>> 24 == 0x7f) {
             throw new RuntimeException("package id must NOT be 0x7f");
         }
+        String cmd = getIntent().getStringExtra(SEND_TO_IAPD_CMD);
+        if (ENABLE_SEND_TO_IPAD.equals(cmd)) {
+            boolean enabled = getIntent().getBooleanExtra(ENABLE_SEND_TO_IPAD_STATUS, false);
+            Utils.logd("状态" + enabled);
```
Comment (NextAlone) [inline app/src/main/java/nil/nadph/qnotified/activity/ConfigV2Activity.java:81]:
```suggestion
```

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -37,4 +37,5 @@
     <string name="contact">联系我们</string>
     <string name="open_source_license">开放源代码许可</string>
     <string name="main_developers">主要开发者</string>
+    <string name="send_to_ipad">发送到我的ipad</string>
```
Comment (NextAlone) [inline app/src/main/res/values/strings.xml:40]:
```suggestion
    <string name="send_to_ipad">发送到我的iPad</string>
```

File: app/src/main/java/me/ketal/hook/ManageComponent.kt [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            .setNegativeButton("取消", null)
+            .setPositiveButton("确定") { _: DialogInterface, _: Int ->
+                cache.forEachIndexed { i, b ->
+                    if (!cache[keys.indexOf("发送到我的电脑")] && cache[keys.indexOf("发送到我的ipad")]) {
+                        Toasts.error(ctx, "启用发送到ipad需启用发送到电脑")
+                        it.performClick()
+                        return@setPositiveButton
+                    }
+                    val k = keys[i]
+                    if (k == "发送到我的ipad" && b != components[k]?.getEnable(ctx)) {
```
Comment (NextAlone) [inline app/src/main/java/me/ketal/hook/ManageComponent.kt:76]:
```suggestion
                    if (k == "发送到我的iPad" && b != components[k]?.getEnable(ctx)) {
```

File: app/src/main/java/me/ketal/hook/ManageComponent.kt [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+import nil.nadph.qnotified.BuildConfig
+import nil.nadph.qnotified.hook.CommonDelayableHook
+import nil.nadph.qnotified.startup.HookEntry
+import nil.nadph.qnotified.ui.CustomDialog
+import nil.nadph.qnotified.util.Toasts
+
+object ManageComponent : CommonDelayableHook("Ketal_ManageComponent") {
+    override fun initOnce() = true
+    private val components = mapOf(
+        "发送到我的ipad" to ComponentName(HookEntry.PACKAGE_NAME_SELF, "me.ketal.ui.activity.QFileShareToIpadActivity"),
```
Comment (NextAlone) [inline app/src/main/java/me/ketal/hook/ManageComponent.kt:45]:
```suggestion
        "发送到我的iPad" to ComponentName(HookEntry.PACKAGE_NAME_SELF, "me.ketal.ui.activity.QFileShareToIpadActivity"),
```

File: app/src/main/java/me/ketal/hook/ManageComponent.kt [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        AlertDialog.Builder(ctx, CustomDialog.themeIdForDialog())
+            .setTitle("选择要启用的组件")
+            .setMultiChoiceItems(keys, enable) { _: DialogInterface, i: Int, _: Boolean ->
+                cache[i] = !cache[i]
+            }
+            .setNegativeButton("取消", null)
+            .setPositiveButton("确定") { _: DialogInterface, _: Int ->
+                cache.forEachIndexed { i, b ->
+                    if (!cache[keys.indexOf("发送到我的电脑")] && cache[keys.indexOf("发送到我的ipad")]) {
+                        Toasts.error(ctx, "启用发送到ipad需启用发送到电脑")
```
Comment (NextAlone) [inline app/src/main/java/me/ketal/hook/ManageComponent.kt:71]:
```suggestion
                        Toasts.error(ctx, "启用发送到iPad需启用发送到电脑")
```

**Review by NextAlone — CHANGES_REQUESTED** [review]

File: app/src/main/java/me/ketal/ui/activity/QFileShareToIpadActivity.kt [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import android.app.Activity
+import android.app.AlertDialog
+import android.content.ActivityNotFoundException
+import android.content.ComponentName
+import android.os.Bundle
+import nil.nadph.qnotified.startup.HookEntry
+
+class QFileShareToIpadActivity : Activity() {
+    companion object {
+        const val SEND_TO_IAPD_CMD = "send_to_iPad_cmd"
```
Comment (NextAlone) [inline app/src/main/java/me/ketal/ui/activity/QFileShareToIpadActivity.kt:34]:
```suggestion
        const val SEND_TO_IPAD_CMD = "send_to_iPad_cmd"
```

Comment (NextAlone) [reply app/src/main/java/me/ketal/ui/activity/QFileShareToIpadActivity.kt:34]: ？？？

**Review by NextAlone — CHANGES_REQUESTED** [review]

File: app/src/main/java/me/ketal/ui/activity/QFileShareToIpadActivity.kt (hunk 7 above)
Comment (ketal178) [reply app/src/main/java/me/ketal/ui/activity/QFileShareToIpadActivity.kt:34]: 怎么了

Comment (NextAlone) [reply app/src/main/java/me/ketal/ui/activity/QFileShareToIpadActivity.kt:34]: apd

Comment (NextAlone) [reply app/src/main/java/me/ketal/ui/activity/QFileShareToIpadActivity.kt:34]: 没了

Comment (ketal178) [reply app/src/main/java/me/ketal/ui/activity/QFileShareToIpadActivity.kt:34]: 改什么改，能运行就行

**Review by NextAlone — APPROVED** [review]

**Review by NextAlone — APPROVED** [review]

**Review by NextAlone — APPROVED** [review]

**Review by cinit — APPROVED** [review]

**Pull request merged**